spring-ai-core/src/main/java/org/springframework/ai/model/security/ApiKey.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2023-2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.ai.model.security;
+
+/**
+ * Some model providers API leverage short-lived api keys which must be renewed at regular
+ * intervals using another credential. For example, a GCP service account can be exchanged
+ * for an api key to call Vertex AI.
+ *
+ * Model clients use the ApiKey interface to get an api key before they make any request
+ * to the model provider. Implementations of this interface can cache the api key and
+ * perform a key refresh when it is required.
+ *
+ * @author Adib Saikali
+ */
+public interface ApiKey {
```

Comment:
this should likely extend Supplier

Reply:
for models called using http they will typically use the an `Authorization:` header which should have a string value, and if we want to support none string value we can add something like `getValueAsXYZ` to the `ApiKey` interface. I can see the `ApiKey` interface with extra methods to get metadata about keys, so 100% agree we should have a rich type to represent a key.  

The `ApiKey` interface is the rich interface I did not want to call it `Key` because that name is too generic and has multiple meaning in the security context for example encryption key ... etc.  

I think using Supplier<ApiKey> makes a lot of sense, I except something like this would be needed in the auto configuration layer. However, I have not touched the auto configuration layer, to keep the scope of the PR small. 

I thought about using `Supplier<ApiKey>` in the `OpenAiApi` class  

Instead of 
```
public class OpenAiApi {
	public OpenAiApi(ApiKey apiKey) {
		this(OpenAiApiConstants.DEFAULT_BASE_URL, apiKey);
	}
```

change to something like 

```java 
public class OpenAiApi {
	public OpenAiApi(Supplier<ApiKey> apiKey) {
		this(OpenAiApiConstants.DEFAULT_BASE_URL, apiKey);
	}
```

The downside of making such a change is that it makes the call sites where `OpenAiApi` is instantiated in numerous test cases full of boiler plate around Supplier. We have a lot of lines code that look like this 

```
OpenAiApi openAiApi = new OpenAiApi(new StaticApiKey(System.getenv("OPENAI_API_KEY")));
````

if we change 

```java
public interface ApiKey {

	/**
	 * Returns an api key to use for a making request. Users of this method should NOT
	 * cache the returned api key, instead call this method whenever you need an api key.
	 * Implementors of this method MUST ensure that the returned key is not expired.
	 * @return the current value of the api key
	 */
	String getValue();
}
```

to 
```java
public interface ApiKey extends Supplier<String> {
}
```

then we can get rid of the `getValue` method since Supplier already has a method called `value`. but then we loose the javadoc defining the contract on the `getValue` method 

Let me know how you want me to proceed.